? custom_url_rewrite_outbound.patch
? multiple_node_access.patch
Index: includes/common.inc
===================================================================
RCS file: /cvs/drupal/drupal/includes/common.inc,v
retrieving revision 1.611.2.24
diff -u -p -r1.611.2.24 common.inc
--- includes/common.inc	13 May 2009 17:10:36 -0000	1.611.2.24
+++ includes/common.inc	31 May 2009 18:19:47 -0000
@@ -1320,7 +1320,17 @@ function url($path = NULL, $query = NULL
     $clean_url = (bool)variable_get('clean_url', '0');
   }
 
-  $base = ($absolute ? $base_url . '/' : base_path());
+  // Preserve the original path before aliasing.
+  $original_path = $path;
+  // Prevent the global $base_url from being altered.
+  $current_base_url = $base_url;
+  
+  if (function_exists('custom_url_rewrite_outbound')) {
+    // Modules may alter outbound links by reference.
+    custom_url_rewrite_outbound($path, $query, $fragment, $absolute, $current_base_url, $original_path);
+  }
+
+  $base = ($absolute ? $current_base_url . '/' : base_path());
 
   // The special path '<front>' links to the default front page.
   if (!empty($path) && $path != '<front>') {
