--- a/domain_config/src/DomainConfigOverrider.php
+++ b/domain_config/src/DomainConfigOverrider.php
@@ -12,6 +12,7 @@ use Drupal\Core\Config\StorageInterface;
 use Drupal\Core\Config\ConfigFactoryOverrideInterface;
 use Drupal\domain\DomainNegotiatorInterface;
 use Drupal\domain\DomainInterface;
+use Drupal\Core\Cache\CacheableMetadata;

 /**
  * Domain-specific config overrides.
@@ -117,5 +118,13 @@ class DomainConfigOverrider implements ConfigFactoryOverrid
     return NULL;
   }

+  /**
+   * {@inheritdoc}
+   * Fixed
+   */
+  public function getCacheableMetadata($name) {
+    $metadata = new CacheableMetadata();
+    return $metadata;
+  }
 }
